### ISSUE TITLE
[PATCH]Multiple header with same name support in http

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7478]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#165|
|Type|defect|
|Created|2003-08-03 21:07:05Z|
|Branch|https://github.com/twisted/twisted/tree/web-headers-165-2|

Attachments:

* [filefile](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/74c/74cbd2c215c2c13c4b6110ada96de8891b355dda/c90ec4ab6fea6e24ab4bb458503403a55c79831c) (1184 bytes) - added by moshez on 2003-10-20 19:06:50Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__165 165
type__defect defect
reporter__itamarst itamarst
priority__high high
milestone__ 
branch__branches_web_headers_165_2 branches/web-headers-165-2
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1059944825000000 1059944825000000
changetime__1212976132000000 1212976132000000
version__ 
owner__ 
cc__glyph cc__itamarst cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
HTTP currently assumes only one header with same name. this
is wrong. Howeve, HTTP specs says this type of header should
only be those headers where it can be replaced with a single
header separated by ";", IIRC. ("Foo: bar; baz;" instead of
two Foo headers.) So we can fake it out that way, or instead
have a new interface that is more lenient than the http rfc.

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
dum di dum

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)|@moshez commented|
|-|-|

```
#!html
<pre>
Here's a patch to solve this.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
Um, my copy of the RFC (http://www.w3.org/Protocols/rfc2616/rfc2616.html) says you 
should use a comma, not a semicolon-space:
"""
Multiple message-header fields with the same field-name MAY be present in a message if and only 
if the entire field-value for that header field is defined as a comma-separated list [i.e., #(values)]. It 
MUST be possible to combine the multiple header fields into one "field-name: field-value" pair, 
without changing the semantics of the message, by appending each subsequent field-value to the 
first, each separated by a comma. The order in which header fields with the same field-name are 
received is therefore significant to the interpretation of the combined field value, and thus a proxy 
MUST NOT change the order of these field values when a message is forwarded.
"""
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
Also, cookie parsing (t.p.http.parseCookies) should be patched to accept both ';' and ',' as 
delimiters, so that multiple Cookie: headers will get parsed properly after being combined here. 
This is also condoned by the RFC:
http://www.w3.org/Protocols/rfc2109/rfc2109.txt
"""
   Note: For backward compatibility, the separator in the Cookie header
   is semi-colon (;) everywhere.  A server should also accept comma (,)
   as the separator between cookie-values for future compatibility.
"""
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
Also, t.w.client currently does support multiple headers via representing them as a list. This is a 
bad idea, because now clients must support both multiple items in the list, and one item in the list 
separated by commas. It is likely that people will get this wrong. HTTPClientFactory does.

Note that it is impossible to automatically un-comma-ify headers into lists, because some headers 
are not allowed to be specified multiple times and thus may contain commas in their data (e.g. all 
the headers that specify dates).

Thus, the best thing to do is to automatically comma-ify multiple instances of the header, and 
make users of specific headers where it's valid split on ','.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
Ugh, I just noticed a ugly problem...the "old" cookie spec (aka, what everyone actually uses) has 
values for Set-Cookie that include an embedded comma. So, HTTPClient is done the way it 
is (against RFC) in order to work in the real world...

Perhaps there is some ugly workaround that could be done to deal with old-style Set-Cookie 
headers (perhaps transform them into new-style Set-Cookie headers?) in such a way as to still be 
RFC compliant with all other headers.

"""
   Netscape's original proposal defined an Expires header that took a
   date value in a fixed-length variant format in place of Max-Age:

   Wdy, DD-Mon-YY HH:MM:SS GMT

   Note that the Expires date format contains embedded spaces, and that
   "old" cookies did not have quotes around values.  Clients that
   implement to this specification should be aware of "old" cookies and
   Expires.
"""
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
And lets just keep writing here...looks like split(',') is not correct for separating headers, as you 
need to take into account quoted strings, which are allowed to include commas.

E.g seeing as incoming headers:
Pragma: x="foo, bar"
Pragma: y="wassup, doc?"
Pragma: z=yup, zz=really
is equivilent to:
Pragma: x="foo, bar", y="wassup, doc?", z=yup, zz=really

and both should be treated as 4 (not 6 or 3) pragma options (x, y, z, and zz).
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
I'm going to punt on the unification and just allow support for multiple
headers. Anything else risks breaking someone's broken-ass http client. Guess
all this crap goes into my protocol design article series.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
Still need to support incoming headers (on both client and server) that contain comma-separated 
values. 

How's this for an interface: make two new functions on http.request: 
getHeaderList(self, key): Call this to retrieve headers which may be specified multiple times. This 
function will return a list of all headers with name 'key', after doing comma-separation. *NOTE*: do 
not use this function to retrieve the "Set-Cookie" header, as the spec for that is broken. Use 
"getRawHeaderList" instead.

getRawHeaderList(self, key): returns the raw list of headers given without doing comma-
separation. Should usually only be used to retrieve the "Set-Cookie" header.

This would be in addition to the existing:
getHeader(self, key)": return the first (or should it be last?) header which matches key.

Possibly there also should be an "appendHeader(self, k, v)" call in addition to "setHeader", for those 
headers which can be specified multiple times. (setHeader would replace all mentions of that 
header, while appendHeader would add a second mention of the header).
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
Okay, I looked up what apache does, cause I figure if twisted does the same thing as it, can't go 
wrong. :)

What it seems to do is ", " merge all headers of the same key. This looks like it is done for 
*all* headers coming into and going out of the server-proper of apache. In the CGIscript->HTTP 
response code path, it has code that will extract all "Set-Cookie" headers before the merging 
happens, and then reinsert them afterwards.

In the HTTP proxy code, it passes headers through as they arrived.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

```
#!html
<pre>
Not to discourage future contributions here, but apparently you're the most
interested in this :).  Mind taking it over and writing a short patch or two?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
Fixed for new-web branch.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `closed`|
|-|-|

Web is dead.  Long live Web2.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set status to `reopened`|
|-|-|

I plead temporary insanity.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

(In [22937]) Branching to 'web-headers-165'

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_17"></a>|@dreid commented|
|-|-|

(In [22980]) Add minimal Headers object implementation to twisted.web from twisted.web2.
Currently only supports raw headers but does support multivalued headers and
overriding name canonicalization.

Refs [#7478](https://github.com/twisted/twisted/issues/7478)

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_18"></a>|@dreid commented|
|-|-|

(In [23126]) Fix a bunch of failures and cause some more.

Refs [#7478](https://github.com/twisted/twisted/issues/7478)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun commented|
|-|-|

(In [23756]) Branching to 'web-headers-165-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Seems to be in reasonable shape now.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_21"></a>|@therve set owner to @exarkun|
|-|-|

 * Funny typo in test_http.ChannelTests.test_headerReceivedMultiple docstring. I let you read that one :)
 * test_http_headers is missing a copyright notice
 * Headers.__getitem__ is not tested with a missing key
 * in hasHeader, removeHeader and setRawHeaders, if I remove the lower() calls nothing breaks. I don't think there are tests with capitalized header names.
 * is the dict API for backward compatibility complete enough? It seems that at least items() and values() are missing.

Thanks

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_22"></a>|@therve commented|
|-|-|

The python2.3 builder seems unhappy: http://buildbot.twistedmatrix.com/builders/debian-py2.3-select/builds/2707/steps/trial/logs/problems.

I think you're using set instead of twisted.python.compat.set.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun commented|
|-|-|

(In [23768]) address review comments

refs [#7478](https://github.com/twisted/twisted/issues/7478)

  * fix typo in test_headerReceivedMultiple docstring
  * add copyright notice and docstring to test_http_headers
  * test Header.__getitem__ raising KeyError
  * add assertions to depend on .lower() calls in Headers
  * add much more of the dict interface, and tests
  * remove deprecations because deprecated() does not work on methods

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun removed owner|
|-|-|

I switched to `twisted.python.compat.set` as well.

Thanks for the review, please have another look. :)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_25"></a>|@therve set owner to @exarkun|
|-|-|

 * ```twisted/web/test/test_http_headers.py:10: '_unreleasedVersion' imported but unused```. I guess that means you can remove some stuff from http_headers as well.
 * the test `HeaderDictMixinTests` should probably be renamed
 * 2.3 buildbot is still not happy: test_http should use compat as well. The more troublesome one is this:
```
[ERROR]: twisted.web.test.test_http_headers.HeaderDictMixinTests.test_update

Traceback (most recent call last):
  File "/srv/bb-slave/Run/slave/full2.3/Twisted/twisted/web/test/test_http_headers.py", line 418, in test_update
    h.update({"foo": "panda", "bar": "marmot"}, baz="lemur")
  File "/srv/bb-slave/Run/slave/full2.3/Twisted/twisted/web/http_headers.py", line 289, in update
    helper.update(*args, **kw)
exceptions.TypeError: update() takes no keyword arguments
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_26"></a>|@exarkun commented|
|-|-|

(In [23778]) respond to review feedback

  * use t.p.c.set in test_http as well
  * get rid of unused deprecation stuff in http_headers
  * rename dict-api Headers test case, since DictMixin is no longer used

refs [#7478](https://github.com/twisted/twisted/issues/7478)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_27"></a>|@exarkun commented|
|-|-|

(In [23779]) Remove some more unused deprecation stuff
refs [#7478](https://github.com/twisted/twisted/issues/7478)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_28"></a>|@exarkun commented|
|-|-|

(In [23780]) skip dict keyword test on python 2.3
refs [#7478](https://github.com/twisted/twisted/issues/7478)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_29"></a>|@exarkun removed owner|
|-|-|

Okay, hopefully better now.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_30"></a>|@jyknight commented|
|-|-|

Is it a good idea to for this object to have the dict interface for raw values? It seems a bit confusing. What will the API for cooked values be?

I seem to recall seeing some code in the past that did things like req.headers = {'foo': 'bar'}. This seems like it probably will break that.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_31"></a>|@therve set owner to @exarkun|
|-|-|

The branch looks ok. I've spotted another problem with nevow tests suite: it compares the headers with a dict, so that fails. Maybe `Headers` should get a `__eq__` too.

I don't know what to think about `req.headers = {'foo': 'bar'}`.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_32"></a>|@exarkun commented|
|-|-|

Replying to [jknight](#note_30):
> Is it a good idea to for this object to have the dict interface for raw values? It seems a bit confusing. What will the API for cooked values be?

Maybe it isn't.  This is the most convenient thing to implement for now.  The idea is that the dict API is deprecated (although the code doesn't reflect this at all right now - I'm just trying to provide the necessary compatibility).  It has to work this way because the old API was dict-of-raw-values.

>
> I seem to recall seeing some code in the past that did things like req.headers = {'foo': 'bar'}. This seems like it probably will break that.

Indeed.  I suppose that case should be supported, as well as comparison as therve pointed out.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_33"></a>|@exarkun commented|
|-|-|

(In [23787]) implement comparison between Headers instances and also dicts

refs [#7478](https://github.com/twisted/twisted/issues/7478)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_34"></a>|@exarkun commented|
|-|-|

(In [23788]) support assignment of dicts to received_headers

refs [#7478](https://github.com/twisted/twisted/issues/7478)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_35"></a>|@exarkun commented|
|-|-|

(In [23789]) split backwards compatibility API into a separate object; expose two APIs at different attributes of Request

refs [#7478](https://github.com/twisted/twisted/issues/7478)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_36"></a>|@exarkun removed owner|
|-|-|

Okay.  Lots more compatibility now.  I even reverted some of the changes in other parts of twisted.web so that they're still using the old API and stuff works.

A couple possible issues:

  * should I deprecate the backwards compatibility APIs now?  This is old stuff.  I definitely want to use `PendingDeprecationError` here.  I guess that means I can't use `deprecated`.  I probably can't anyway, since the deprecations need to happen in certain cases in `__setattr__` and... maybe `__getattr__`, urg.
  * if there were multiple _content-type_ or _connection_ headers before, the last value was used.  now the first is.  I dunno if that is important and/or good/bad.

A nice thing is that `Headers` now doesn't have any backwards compat junk on it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_37"></a>|@exarkun commented|
|-|-|

(In [23794]) fix trivially invalid epytext markup

refs [#7478](https://github.com/twisted/twisted/issues/7478)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_38"></a>|@therve set owner to @exarkun|
|-|-|

 * _DictHeaders mentions received_headers, it could mention headers too
 * typo in DictHeaders.get docstring: "is not present/"
 * same in pop
 * RequestTests._compatHeadersTest should have a docstring

Deprecation would be nice, but I don't think it has to happen right now. We should first improve our deprecation system anyway.

I'm +1 for merging once the trivial points are done.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_39"></a>|@exarkun commented|
|-|-|

(In [23809]) fix docstrings and add one which was missing

refs [#7478](https://github.com/twisted/twisted/issues/7478)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_40"></a>|@exarkun commented|
|-|-|

(In [23825]) adjusterations

refs [#7478](https://github.com/twisted/twisted/issues/7478)

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_41"></a>|@radix commented|
|-|-|

+1 on LICENSE changes.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_42"></a>|@exarkun set status to `closed`|
|-|-|

(In [23826]) Merge web-headers-165-2

Author: dreid, exarkun
Reviewer: therve
Fixes: [#7478](https://github.com/twisted/twisted/issues/7478)

Add two new attributes to twisted.web.http.Request, `requestHeaders` and
`responseHeaders`, which parallel the previous attributes `received_headers`
and `headers`.  The new attributes are preferred and provide a structured
interface to accessing and manipulating header information, including
multiples values per header which was unsupported by the previous API.
`received_headers` and `headers` are not deprecated and coding using
them will continue to work, but their use is discouraged and they will be
deprecated at some future time.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_43"></a>|@exarkun set status to `reopened`|
|-|-|

(In [23891]) Revert r23826 - backward incompatibility in received_header iteration

Reopens: [#7478](https://github.com/twisted/twisted/issues/7478)

With this changeset, it is no longer possible to iterate over `Request.received_headers`
or `Request.headers`.  This is because `_DictHeaders` does not implement `__iter__` and
the fallback logic in Python for iteration only works for sequences, not mappings.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_44"></a>|@exarkun commented|
|-|-|

(In [23892]) implement `__iter__` for `_DictHeaders`

refs [#7478](https://github.com/twisted/twisted/issues/7478)

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_45"></a>|@jyknight commented|
|-|-|

_DictHeaders should certainly be a subclass of UserDict.DictMixin. DictMixin has a well-defined set of required and optional-for-higher-performance methods to implement upon which it implements the entire dict api.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_46"></a>|@exarkun commented|
|-|-|

(In [23893]) fix typo; implement `__len__`

refs [#7478](https://github.com/twisted/twisted/issues/7478)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_47"></a>|@exarkun commented|
|-|-|

(In [23894]) implement setdefault, too

refs [#7478](https://github.com/twisted/twisted/issues/7478)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_48"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

The code's written and tested now.  I don't want to use `DictMixin` because we don't control it.  python-dev has developed in me a fear of using anything like this from the stdlib.

I've added `__len__`, `__iter__`, and `setdefault`.  I don't think anything else that we need to provide is missing:

```python
>>> from twisted.web.http_headers import _DictHeaders
>>> set(dir(dict)) - set(dir(_DictHeaders))
set(['__ne__', 'fromkeys', '__le__', '__gt__', '__lt__', '__eq__', '__ge__'])
>>>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_49"></a>|@exarkun commented|
|-|-|

(In [23895]) Get rid of as much of the dict implementation as possible, in favor of `DictMixin`

refs [#7478](https://github.com/twisted/twisted/issues/7478)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_50"></a>|@exarkun commented|
|-|-|

Okay.  foom convinced me.

**Comment by twisted-trac** [comment]
|<img alt="pjd's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_51"></a>|pjd commented|
|-|-|

Note:  This branch contains a sneaky regression in `HTTPChannel.headerReceived`:  `self.maxHeaders` is no longer compared to the current request's header count, but to an instance-wide count which is never reset:  multiple normal requests keep incrementing it until it hits `maxHeaders`, resulting in spurious (and really hard to debug) `400 Bad Request` errors.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_53"></a>|@exarkun commented|
|-|-|

(In [23898]) get rid of duplication in `ChannelTests`, move tests to `ParsingTestCase`, add some tests for multiple requests, fix the bug

refs [#7478](https://github.com/twisted/twisted/issues/7478)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_54"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

The max header issue should be resolved by the previous commit.  I took the opportunity to clean up the tests slightly as well.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_55"></a>|@exarkun commented|
|-|-|

(In [23901]) add a custom setdefault implementation since Python 2.3's DictMixin has a signature incompatible with dict.setdefault

refs [#7478](https://github.com/twisted/twisted/issues/7478)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_56"></a>|@therve set owner to @exarkun|
|-|-|

Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_57"></a>|@exarkun set status to `closed`|
|-|-|

(In [23915]) Merge web-headers-165-2

Author: dreid, exarkun
Reviewer: therve
Fixes: [#7478](https://github.com/twisted/twisted/issues/7478)

Add two new attributes to twisted.web.http.Request, requestHeaders
and responseHeaders, which parallel the previous attributes
received_headers and headers. The new attributes are preferred and
provide a structured interface to accessing and manipulating header
information, including multiples values per header which was
unsupported by the previous API. received_headers and headers are
not deprecated and code using them will continue to work, but their
use is discouraged and they will be deprecated at some future
time.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
